### PR TITLE
Module-level static declarations (closes #7452)

### LINKED ISSUE #7452: Module level functions
https://github.com/HaxeFoundation/haxe-evolution/pull/24

### PR DESCRIPTION
**Pull request by nadako**
This PR implements module-level function, var and property declaration (proposed and discussed in https://github.com/HaxeFoundation/haxe-evolution/pull/24):

```haxe
final version = 4;

function main() {
    trace("Hello from version " + version);
}
```

<details>
<summary>Generated JS</summary>

```js
function Main_main() {
	console.log("src/Main.hx:4:","Hello from version " + Main_version);
}
var Main_version = 4;
Main_main();
```
</details>

This is implemented the following way:
 * Definition:
   * Parse such declarations into the new `type_def` variant: `EStatic of (placed_access, class_field_kind) definition`
   * When loading module, all its `EStatic`s are collected into a list of `static` fields for an implicitly created class (fields are `public` unless explicitly defined as `private`, just like types)
   * Class is created with a special `cl_kind`:  `KModuleStatics of module_def`
   * Class is then stored into its module using new field `m_statics : tclass option`
 * Import:
   * `import Mod` will import all the module-level statics
   * `import Mod.func [as name]` will first try importing module-level `func` and `Mod.func` static field
   * `import Mod.*` will NOT import module-level statics (debatable, but consistent with it not importing other module-level declarations)
 * Resolution:
   * Unqualified name will first search for this-module statics and then imported ones
   * Qualified path will first look for module-level statics and then fall back to sub-type and main-type-static resolution.
 * Misc:
   * `-main Mod` works with module-level `main` function
   * `using Mod` will consider module-level static extension functions
   * Module-level `macro` functions are also supported
 * Macro:
   * New `haxe.macro.Type.ClassKind` variant: `KModuleStatics(module:String)`
   * New `haxe.macro.Expr.TypeDefKind` variant: `TDStatic(kind:FieldType, ?access:Array<Access>)`
 * Generation:
   * Target generator can choose to not generate `KModuleStatics` class and instead output functions and vars directly (implemented for JS)
   * If target still has to wrap them in class it can at least omit reflection helpers and apply extra optimizations (e.g. add `static`/`final` modifiers)

TODO checklist:

 - [x] handle imported module-statics resolution
 - [x] handle `import pack.Module.moduleLevel`
 - [x] handle `import pack.Module.moduleLevel as otherName`
 - [x] handle `pack.Module.moduleLevel` access (the MORDOR)
 - [x] refactor TyperDotPath into something readable
 - [x] finish parsing, support var, property, modifiers, etc.
 - [x] adjust haxe.macro.Expr API
 - [x] think of macro API to retrieve the statics class
 - [x] discuss and decide module-scoping or package scoping
 - [x] think of static extensions
 - [ ] handle display stuff
   - [ ] `resume`
 - [x] double-check positions
 - [ ] check access control (i.e. `@:access`)
 - [ ] check `@:rtti` (ugh)
 - [x] check `haxe.PosInfos` 
   - (not) solve this as part of https://github.com/HaxeFoundation/haxe/issues/9214
 - [x] add tests
   - [x] test module-level -main on different targets (e.g. java)
   - [x] test `Context.defineType/defineModule` for statics
 - [x] support class-less `main`
 - [x] support class-less `--macro` calls and `@:build` (tame eval)
 - [x] flatten module-statics class in genjs (as an example)
 - [x] discuss `import Mod.*`
 - [x] filter out `private` when importing not to pollute the local namespace (prevents `@:privateAccess` but is consistent with private types)
--- 
Things that can be done after the merge: https://github.com/HaxeFoundation/haxe/issues/9436

### CODE REVIEW

**Comment by RealyUniqueName** [comment]
Why is it called "globals" in AST but "module statics" after typing?

**Comment by nadako** [comment]
> Why is it called "globals" in AST but "module statics" after typing?

Naming is hard :)

Well the idea is that the untyped AST does not care about whether it's implemented as statics or not, so it uses the term "global", which might as well be "module-local" or something, I'm not really sure.

In the typed AST it's implemented as statics collected in a class, so it makes sense to name it as that.

File: src/typing/typeloadModule.ml [hunk 1]
```diff
@@ -545,13 +545,28 @@ let init_module_type ctx context_init (decl,p) =
 						ctx.m.module_types <- ((match name with None -> tsub | Some(n,pname) -> rebind tsub n pname),p) :: ctx.m.module_types
 					with Not_found ->
 						(* this might be a static property, wait later to check *)
-						let tmain = get_type tname in
-						context_init#add (fun() ->
+						let find_main_type_static () = 
+							let tmain = get_type tname in
```
Comment (nadako) [inline src/typing/typeloadModule.ml:566]: @Simn is it okay to delay this `get_type` to whenever `context_init#add` callback is called?

**Comment by nadako** [comment]
> Well the idea is that the untyped AST does not care about whether it's implemented as statics or not, so it uses the term "global", which might as well be "module-local" or something, I'm not really sure.

Thinking about this with a "fresh" head, maybe we should officially call them module statics. It is easy to pronounce, close to the implementation and rather easy to understand: "like class statics, but module statics".

**Comment by nanjizal** [comment]
So you could have module level static functions, with module level properties and constants ( finals ) soon to be implemented, but are you going to change EGlobal to EModule ?

**Comment by nadako** [comment]
EStatic, I guess, but yeah :)

**Comment by fullofcaffeine** [comment]
Will modules have some kind of initialiser hook? I mean, in js you can just add code to the body to be ran when at runtime when the module is interpreted and it can be quite useful.

**Comment by nadako** [comment]
> Will modules have some kind of initialiser hook? I mean, in js you can just add code to the body to be ran when at runtime when the module is interpreted and it can be quite useful.

I am not sure what exactly are you asking here. In JS modules are loaded at run-time, while in Haxe we just load them at compile-time to find and build the classes, there's no concept of a module at run-time.

**Comment by fullofcaffeine** [comment]
I'm aware that it's a compile-time feature. I was just curious if some abstraction for initialization would be provided that would then run at runtime, but I'm probably not fully aware of the implementation and how it's going to work.

**Comment by RealyUniqueName** [comment]
I wonder if defining `function __init__()` would automatically work with the same semantics as `static function __init__()` because module functions get compiled as static function of an auto-genetared class.

**Comment by nadako** [comment]
I think it will, but so will var initializations - both will be executed at program startup.

File: src/typing/typeloadModule.ml [hunk 2]
```diff
@@ -189,8 +190,18 @@ end
...
 	let com = ctx.com in
 	let decls = ref [] in
+	let statics = ref [] in
+	let check_name name p =
+		if
+			List.exists (fun (t,_) -> snd (t_path t) = name) !decls
+			||
+			List.exists (fun (d,_) -> fst d.d_name = name) !statics
+		then
+			(* TODO: should this be just "Name ... is already defined" now? *)
```
Comment (nadako) [inline src/typing/typeloadModule.ml]: Should it?

File: src/typing/typeloadModule.ml [hunk 3]
```diff
@@ -193,12 +193,21 @@ end
...
+			error ("Type name " ^ name ^ " is already defined in this module") p
+	in
 	let make_path name priv =
-		if List.exists (fun (t,_) -> snd (t_path t) = name) !decls then error ("Type name " ^ name ^ " is already defined in this module") loadp;
+		check_name name loadp; (* TODO: I think this should be the definition path, not loadp *)
```
Comment (nadako) [inline src/typing/typeloadModule.ml]: TODO

File: src/typing/typeloadModule.ml (hunk 1 above)
Comment (nadako) [reply src/typing/typeloadModule.ml:566]: oh well, seems to work

**Comment by nadako** [comment]
This should close #7452

**Comment by nadako** [comment]
(blindly) rebased on top development with all the latest changes, let's see if tests still pass

**Review by Simn — APPROVED** [review]

**Comment by nadako** [comment]
I need to check and add test for the behaviour of module-statics regarding #9367, and will merge afterwards.

**Comment by jeremyfa** [comment]
A fun side effect of this evolution is that one could create module level code in, say, `Mod.hx` and make it compatible with hscript very easily (no surrounding class etc...).

Looking awesome overall!

**Comment by uvtc** [comment]
Any guess as to what release this will land in?

**Comment by uvtc** [comment]
Whoops. Sorry, just realized <https://github.com/HaxeFoundation/haxe-evolution/blob/master/proposals/0007-module-level-funcs.md> says 4.2.0.

**Comment by codingthat** [comment]
Just curious, are these still to do, or just didn't get checked off formally?
image

**Comment by nadako** [comment]
Still kinda to do. For the "display stuff" there are separate issues now, rtti and access controls are... also waiting for specific issues :)

**Pull request merged**